File: ucharm/compat/typing.py
# ucharm/compat/typing.py
"""
Stub implementation of typing for MicroPython.

This provides no-op versions of all typing constructs so that
code using type hints can run without modification.

Type hints have no runtime effect in Python, so we just need
to provide the names - they don't need to do anything.
"""


# Special typing primitives
class _SpecialForm:
    """Base for special typing forms."""

    def __init__(self, name):
        self._name = name

    def __repr__(self):
        return "typing." + self._name

    def __getitem__(self, params):
        return self

    def __call__(self, *args, **kwargs):
        # For things like cast()
        if args:
            return args[0]
        return None


# Basic type aliases - these are just identity for actual types
Any = _SpecialForm("Any")
NoReturn = _SpecialForm("NoReturn")
Never = _SpecialForm("Never")

# Generic aliases
Union = _SpecialForm("Union")
Optional = _SpecialForm("Optional")
List = _SpecialForm("List")
Dict = _SpecialForm("Dict")
Set = _SpecialForm("Set")
FrozenSet = _SpecialForm("FrozenSet")
Tuple = _SpecialForm("Tuple")
Type = _SpecialForm("Type")
Callable = _SpecialForm("Callable")
Sequence = _SpecialForm("Sequence")
Mapping = _SpecialForm("Mapping")
MutableMapping = _SpecialForm("MutableMapping")
MutableSequence = _SpecialForm("MutableSequence")
MutableSet = _SpecialForm("MutableSet")
Iterable = _SpecialForm("Iterable")
Iterator = _SpecialForm("Iterator")
Generator = _SpecialForm("Generator")
Coroutine = _SpecialForm("Coroutine")
AsyncGenerator = _SpecialForm("AsyncGenerator")
AsyncIterable = _SpecialForm("AsyncIterable")
AsyncIterator = _SpecialForm("AsyncIterator")
Awaitable = _SpecialForm("Awaitable")
ContextManager = _SpecialForm("ContextManager")
AsyncContextManager = _SpecialForm("AsyncContextManager")
Pattern = _SpecialForm("Pattern")
Match = _SpecialForm("Match")
IO = _SpecialForm("IO")
TextIO = _SpecialForm("TextIO")
BinaryIO = _SpecialForm("BinaryIO")

# Python 3.9+ style generics
Literal = _SpecialForm("Literal")
Final = _SpecialForm("Final")
ClassVar = _SpecialForm("ClassVar")
Annotated = _SpecialForm("Annotated")
TypeGuard = _SpecialForm("TypeGuard")
Concatenate = _SpecialForm("Concatenate")
ParamSpec = _SpecialForm("ParamSpec")
TypeVarTuple = _SpecialForm("TypeVarTuple")
Unpack = _SpecialForm("Unpack")
Self = _SpecialForm("Self")
LiteralString = _SpecialForm("LiteralString")
Required = _SpecialForm("Required")
NotRequired = _SpecialForm("NotRequired")


class TypeVar:
    """Type variable for generic types."""

    def __init__(
        self, name, *constraints, bound=None, covariant=False, contravariant=False
    ):
        self.__name__ = name
        self.__constraints__ = constraints
        self.__bound__ = bound
        self.__covariant__ = covariant
        self.__contravariant__ = contravariant

    def __repr__(self):
        return "~" + self.__name__


class Generic:
    """Base class for generic types."""

    def __class_getitem__(cls, params):
        return cls


class Protocol:
    """Base class for protocol classes."""

    pass


class NamedTuple:
    """Typed version of namedtuple."""

    def __init_subclass__(cls, **kwargs):
        pass


class TypedDict(dict):
    """Dictionary with a fixed set of keys and value types."""

    def __init_subclass__(cls, total=True, **kwargs):
        pass


# Functions
def cast(typ, val):
    """Cast a value to a type (no-op at runtime)."""
    return val


def overload(func):
    """Decorator for overloaded functions (no-op at runtime)."""
    return func


def final(func):
    """Decorator to indicate a final method/class (no-op at runtime)."""
    return func


def no_type_check(func):
    """Decorator to indicate no type checking (no-op at runtime)."""
    return func


def no_type_check_decorator(decorator):
    """Decorator to disable type checking for a decorator."""
    return decorator


def runtime_checkable(cls):
    """Mark a protocol as runtime checkable."""
    return cls


def get_type_hints(obj, globalns=None, localns=None, include_extras=False):
    """Return type hints for an object."""
    hints = getattr(obj, "__annotations__", None)
    return hints if hints else {}


def get_origin(tp):
    """Get the unparameterized version of a type."""
    return getattr(tp, "__origin__", None)


def get_args(tp):
    """Get type arguments of a parameterized type."""
    return getattr(tp, "__args__", ())


def is_typeddict(tp):
    """Check if a type is a TypedDict."""
    return isinstance(tp, type) and issubclass(tp, TypedDict)


# Type aliases that are just the real types
AnyStr = TypeVar("AnyStr", str, bytes)
Text = str

# Collections
Counter = dict
Deque = list
DefaultDict = dict
OrderedDict = dict
ChainMap = dict

# For typing.TYPE_CHECKING
TYPE_CHECKING = False


# NewType creates a distinct type for type checkers
def NewType(name, tp):
    """Create a distinct type for type checking purposes."""

    def new_type(x):
        return x

    new_type.__name__ = name
    new_type.__supertype__ = tp
    return new_type


# reveal_type is used for debugging type inference
def reveal_type(obj):
    """Reveal the inferred type of an expression (for debugging)."""
    return obj


# assert_type is used for type assertions
def assert_type(val, typ):
    """Assert that a value has the expected type."""
    return val


# assert_never is used for exhaustiveness checking
def assert_never(arg):
    """Assert that this code is never reached."""
    raise AssertionError("Expected code to be unreachable")


# dataclass_transform decorator (Python 3.11+)
def dataclass_transform(**kwargs):
    """Decorator for dataclass-like decorators."""

    def decorator(cls_or_fn):
        return cls_or_fn

    return decorator

File: runtime/compat/typing.zig
const pk = @import("pk");
const c = pk.c;

var tp_typevar: c.py_Type = 0;

// Helper to create a simple type alias (a class that acts as a type hint)
fn createTypeAlias(module: c.py_Ref, name: [:0]const u8) void {
    const tp = c.py_newtype(name, c.tp_object, module, null);
    c.py_setdict(module, c.py_name(name), c.py_tpobject(tp));
}

// Helper to set a constant to a sentinel object
fn setTypingConstant(module: c.py_Ref, name: [:0]const u8) void {
    _ = c.py_newobject(c.py_r0(), c.tp_object, -1, 0);
    c.py_setdict(module, c.py_name(name), c.py_r0());
}

// TypeVar.__new__(cls, name, *constraints, bound=None, covariant=False, contravariant=False)
fn typeVarNewFn(ctx: *pk.Context) bool {
    var name_arg = ctx.arg(1) orelse return ctx.typeError("TypeVar() requires a name argument");

    // Create a TypeVar instance
    _ = c.py_newobject(c.py_retval(), tp_typevar, -1, 0);

    // Store the name as __name__ attribute
    c.py_setdict(c.py_retval(), c.py_name("__name__"), name_arg.ref());

    return true;
}

// TypeVar.__repr__
fn typeVarReprFn(ctx: *pk.Context) bool {
    var self = ctx.arg(0) orelse return ctx.typeError("self required");
    const name_val = c.py_getdict(self.ref(), c.py_name("__name__"));
    if (name_val) |n| {
        // Return ~name format
        const sv = c.py_tosv(n);
        var buf: [128]u8 = undefined;
        const prefix = "~";
        @memcpy(buf[0..prefix.len], prefix);
        const name_len: usize = @intCast(sv.size);
        if (name_len + prefix.len < buf.len) {
            @memcpy(buf[prefix.len .. prefix.len + name_len], sv.data[0..name_len]);
            return ctx.returnStr(buf[0 .. prefix.len + name_len]);
        }
    }
    return ctx.returnStrZ("~T");
}

// cast(typ, val) -> val (no-op, just returns the value)
fn castFn(ctx: *pk.Context) bool {
    const val = ctx.arg(1) orelse return ctx.typeError("cast() requires 2 arguments");
    return ctx.returnValue(val);
}

// overload(func) -> func (decorator that returns the function unchanged)
fn overloadFn(ctx: *pk.Context) bool {
    const func = ctx.arg(0) orelse return ctx.typeError("overload() requires 1 argument");
    return ctx.returnValue(func);
}

// final(func) -> func (decorator that returns the function unchanged)
fn finalFn(ctx: *pk.Context) bool {
    const func = ctx.arg(0) orelse return ctx.typeError("final() requires 1 argument");
    return ctx.returnValue(func);
}

// no_type_check(func) -> func (decorator that returns the function unchanged)
fn noTypeCheckFn(ctx: *pk.Context) bool {
    const func = ctx.arg(0) orelse return ctx.typeError("no_type_check() requires 1 argument");
    return ctx.returnValue(func);
}

// runtime_checkable(cls) -> cls (decorator that returns the class unchanged)
fn runtimeCheckableFn(ctx: *pk.Context) bool {
    const cls = ctx.arg(0) orelse return ctx.typeError("runtime_checkable() requires 1 argument");
    return ctx.returnValue(cls);
}

// get_args(tp) -> () (returns empty tuple for now)
fn getArgsFn(_: *pk.Context) bool {
    _ = c.py_newtuple(c.py_retval(), 0);
    return true;
}

// get_origin(tp) -> None (returns None for now)
fn getOriginFn(ctx: *pk.Context) bool {
    return ctx.returnNone();
}

// get_type_hints(obj) -> {} (returns empty dict for now)
fn getTypeHintsFn(ctx: *pk.Context) bool {
    return ctx.returnDict();
}

pub fn register() void {
    var builder = pk.ModuleBuilder.new("typing");

    // TypeVar - special handling to make it callable
    tp_typevar = c.py_newtype("TypeVar", c.tp_object, builder.module, null);
    c.py_bind(c.py_tpobject(tp_typevar), "__new__(cls, name, *constraints, bound=None, covariant=False, contravariant=False)", pk.wrapFn(2, -1, typeVarNewFn));
    c.py_bindmethod(tp_typevar, "__repr__", pk.wrapFn(1, 1, typeVarReprFn));
    c.py_setdict(builder.module, c.py_name("TypeVar"), c.py_tpobject(tp_typevar));

    // Generic types (type aliases)
    createTypeAlias(builder.module, "List");
    createTypeAlias(builder.module, "Dict");
    createTypeAlias(builder.module, "Set");
    createTypeAlias(builder.module, "FrozenSet");
    createTypeAlias(builder.module, "Tuple");
    createTypeAlias(builder.module, "Type");
    createTypeAlias(builder.module, "Callable");
    createTypeAlias(builder.module, "Generic");
    createTypeAlias(builder.module, "Protocol");

    // Abstract base classes
    createTypeAlias(builder.module, "Sequence");
    createTypeAlias(builder.module, "MutableSequence");
    createTypeAlias(builder.module, "Mapping");
    createTypeAlias(builder.module, "MutableMapping");
    createTypeAlias(builder.module, "Iterable");
    createTypeAlias(builder.module, "Iterator");
    createTypeAlias(builder.module, "Generator");
    createTypeAlias(builder.module, "Reversible");
    createTypeAlias(builder.module, "Container");
    createTypeAlias(builder.module, "Collection");
    createTypeAlias(builder.module, "Hashable");
    createTypeAlias(builder.module, "Sized");

    // Async types
    createTypeAlias(builder.module, "Awaitable");
    createTypeAlias(builder.module, "Coroutine");
    createTypeAlias(builder.module, "AsyncGenerator");
    createTypeAlias(builder.module, "AsyncIterator");
    createTypeAlias(builder.module, "AsyncIterable");

    // IO types
    createTypeAlias(builder.module, "IO");
    createTypeAlias(builder.module, "TextIO");
    createTypeAlias(builder.module, "BinaryIO");

    // Special forms (sentinel objects)
    setTypingConstant(builder.module, "Any");
    setTypingConstant(builder.module, "Optional");
    setTypingConstant(builder.module, "Union");
    setTypingConstant(builder.module, "ClassVar");
    setTypingConstant(builder.module, "Final");
    setTypingConstant(builder.module, "Literal");
    setTypingConstant(builder.module, "Annotated");
    setTypingConstant(builder.module, "NoReturn");
    setTypingConstant(builder.module, "Never");
    setTypingConstant(builder.module, "Self");
    setTypingConstant(builder.module, "LiteralString");
    setTypingConstant(builder.module, "TypeAlias");
    setTypingConstant(builder.module, "Concatenate");
    setTypingConstant(builder.module, "ParamSpec");
    setTypingConstant(builder.module, "TypeVarTuple");
    setTypingConstant(builder.module, "Unpack");
    setTypingConstant(builder.module, "Required");
    setTypingConstant(builder.module, "NotRequired");
    setTypingConstant(builder.module, "ReadOnly");

    // TYPE_CHECKING constant (False at runtime)
    _ = builder.constBool("TYPE_CHECKING", false);

    // Functions
    _ = builder
        .funcWrapped("cast", 2, 2, castFn)
        .funcWrapped("overload", 1, 1, overloadFn)
        .funcWrapped("final", 1, 1, finalFn)
        .funcWrapped("no_type_check", 1, 1, noTypeCheckFn)
        .funcWrapped("runtime_checkable", 1, 1, runtimeCheckableFn)
        .funcWrapped("get_args", 1, 1, getArgsFn)
        .funcWrapped("get_origin", 1, 1, getOriginFn)
        .funcWrapped("get_type_hints", 1, 3, getTypeHintsFn);
}
